### PR TITLE
Az parse xd

### PR DESCRIPTION
**Pull request by andjelx**
Fixed issues with xml parsing. Tested on universal-2006-2015-raw.zip

### CODE REVIEW

File: scripts/18-convert2xd.py [hunk 1]
```diff
@@ -33,6 +33,8 @@ def main():
     global args
     parsers = {
         '.xml': [parse_ccxml, parse_uxml],
+        '.xml.1': [parse_ccxml, parse_uxml],
```
Comment (saulpw) [inline scripts/18-convert2xd.py]: Hm, where are these .1 and .2 files coming from?

File: xdfile/ccxml2xd.py [hunk 2]
```diff
@@ -1,38 +1,38 @@
...
-    content = content.replace("</em>", "/}")
-    content = content.replace("<u>", "{_")
-    content = content.replace("</u>", "_}")
-    content = content.replace("<strike>", "{-")
-    content = content.replace("</strike>", "-}")
+    content = data.decode('utf-8', errors='replace')
+    content = escape(content, xml_escape_table)
+    content = re.sub(r'(=["]{2}([^"]+?)["]{2})+',r'="&quot;\2&quot;"', content) # Replace double quotes
+    content = remove_cons_lines(content)
+    # TO BE REMOVED
```
Comment (saulpw) [inline xdfile/ccxml2xd.py]: Please remove unneeded code

File: xdfile/utils.py [hunk 3]
```diff
@@ -45,8 +56,13 @@ def br_with_n(text):
 def get_log():
     return EOL.join(g_logs) + EOL
 
+def log(s, minverbose=0, severity='INFO'):
+    # This can be made in more Python way
+    if severity.lower() == 'warning':
```
Comment (saulpw) [inline xdfile/utils.py]: let's make 'warn' and 'error' functions that call this helper.

File: xdfile/utils.py [hunk 4]
```diff
@@ -487,3 +506,49 @@ def memoizer(*args, **kwargs):
...
+    "</em>": "/}",
+    "<u>": "{_",
+    "</u>": "_}",
+    "<strike>": "{-",
+    "</strike>": "-}",
+    "&" : "&amp;",
+    "<92>" : "&apos;",
+    '"<"' : '"&lt;"',
+    '="" ' : "='' ",
+    '\x05': "'", # ^E seems to be junk
```
Comment (saulpw) [inline xdfile/utils.py]: Is it junk (nothing) or single quote?

File: xdfile/utils.py [hunk 5]
```diff
@@ -487,3 +506,49 @@ def memoizer(*args, **kwargs):
...
+def escape(data, entities={}):
+    """Escape a string of data.
+    based on: xml.sax.saxutils escape()
+    """
+    if entities:
+        data = __dict_replace(data, entities)
+    return data
+
+def remove_cons_lines(text):
+    """ Remove two consequative lines if equal """
```
Comment (saulpw) [inline xdfile/utils.py]: consecutive (sp)

File: xdfile/uxml2xd.py [hunk 6]
```diff
@@ -70,7 +66,11 @@ def parse_uxml(content, filename):
     for clue_type in ('across', 'down'):
         for clue in root.xpath('//crossword/' + clue_type)[0].getchildren():
             number = int(clue.attrib['cn'])
-            text = udecode(clue.attrib['c'].strip())
+            try:
+                text = udecode(clue.attrib['c'].strip())
+            except:
+                print("Clue: %s" % clue)
```
Comment (saulpw) [inline xdfile/uxml2xd.py]: remove debug code

File: xdfile/utils.py (hunk 4 above)
Comment (andjelx) [inline xdfile/utils.py]: It's single quote. I took it as is from your parser.

File: scripts/18-convert2xd.py (hunk 1 above)
Comment (andjelx) [inline scripts/18-convert2xd.py]: Were some on universal/ - fcx2007-01-02.xml.2
Maybe it's a good idea to skip such...

File: xdfile/ccxml2xd.py (hunk 2 above)
Comment (andjelx) [inline xdfile/ccxml2xd.py]: Done

File: xdfile/utils.py (hunk 5 above)
Comment (andjelx) [inline xdfile/utils.py]: Renamed

File: xdfile/utils.py [hunk 7]
```diff
@@ -487,3 +506,49 @@ def memoizer(*args, **kwargs):
...
+    ret = []
+    for l in text.splitlines():
+        if not ret:
+            ret.append(l)
+        elif l != ret[-1]:
+            ret.append(l)
+    return '\n'.join(ret)
```
Comment (andjelx) [inline xdfile/utils.py]: Done

File: xdfile/utils.py (hunk 3 above)
Comment (andjelx) [inline xdfile/utils.py]: Done.
log_warning() and log_error()

File: xdfile/uxml2xd.py (hunk 6 above)
Comment (andjelx) [inline xdfile/uxml2xd.py]: Done.

**Comment by andjelx** [comment]
Caught few bugs while removing debug code. Now works good.

**Pull request merged**